src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,353 @@
+import app.revanced.patcher.util.xml.LazyXMLInputSource
+import app.revanced.patcher.util.xml.scanIdRegistrations
+import com.reandroid.apk.ApkModule
+import com.reandroid.apk.xmlencoder.EncodeException
+import com.reandroid.apk.xmlencoder.EncodeMaterials
+import com.reandroid.apk.xmlencoder.EncodeUtil
+import com.reandroid.archive.InputSource
+import com.reandroid.arsc.chunk.PackageBlock
+import com.reandroid.arsc.chunk.TableBlock
+import com.reandroid.arsc.chunk.xml.AndroidManifestBlock
+import com.reandroid.arsc.value.Entry
+import com.reandroid.arsc.value.ResConfig
+import lanchon.multidexlib2.BasicDexEntry
+import lanchon.multidexlib2.DexIO
+import lanchon.multidexlib2.MultiDexContainerBackedDexFile
+import lanchon.multidexlib2.MultiDexIO
+import lanchon.multidexlib2.RawDexIO
+import org.jf.dexlib2.dexbacked.DexBackedDexFile
+import org.jf.dexlib2.iface.MultiDexContainer
+import org.jf.dexlib2.writer.io.MemoryDataStore
+import org.w3c.dom.*
+import java.io.File
+import java.util.zip.ZipEntry
+
+
+/**
+ * The [Apk] file that is to be patched.
+ *
+ * @param path The path to the apk file.
+ * @param name The name of this apk.
+ */
+sealed class Apk private constructor(internal val module: ApkModule) {
+    companion object {
+        const val manifest = "AndroidManifest.xml"
+    }
+
+    /**
+     * The metadata of the [Apk].
+     */
+    val packageMetadata = PackageMetadata(module.androidManifestBlock)
+
+    internal val archive = Archive(module)
+
+
+    /**
+     * Refresh updated resources for an [Apk].
+     *
+     * @param options The [PatcherOptions] of the [Patcher].
+     */
+    internal open fun finalize(options: PatcherOptions) {
+        archive.openFiles.forEach { options.logger.warn("File $it was never closed! File modifications will not be applied if you do not close them.") }
+
+        resources.useMaterials {
+            module.apkArchive.listInputSources().forEach {
+                val pkg = resources.packageBlock
+                if (it is LazyXMLInputSource) {
+                    if (resources.hasResourceTable) {
+                        // Scan for @+id registrations.
+                        it.document.scanIdRegistrations().forEach { attr ->
+                            val name = attr.value.split('/').last()
+                            options.logger.trace("Registering ID: $name")
+                            pkg!!.getOrCreate("", "id", name).resValue.valueAsBoolean = false
+                            attr.value = "@id/$name"
+                        }
+                    }
+
+                    try {
+                        it.encode() // Encode the LazyXMLInputSource.
+                    } catch (e: EncodeException) {
+                        throw ApkException.Encode("Failed to encode ${it.name}", e)
+                    }
+                }
+
+                if (it.name == manifest) {
+                    // Update package block name
+                    resources.packageBlock?.name =
```

Comment:
Some apk files such as base or assets might always have a packageBlock. Override finalize for each apk type. This way you can skip the null check

Reply:
Wouldn't the Kotlin compiler still have to emit a null check because `Resources.packageBlock` is still nullable?

Reply:
You can use `!!` to assert its not null.

Reply:
> You can use `!!` to assert its not null.

Yes, but what is the benefit of doing that compared to using the null-safe operator?
If we wanted to make sure all non-library apk files have resource tables then it would be better to just do that inside an `init` block.

Reply:
The reason would be to convey that base or asset always have packageBlock unlike other types of apks; that's the assumption for this review.

Reply:
As I said, there is a better way to convey that, but whatever.